### ISSUE TITLE
GET request validated scheme body?

### ISSUE THREAD

**Issue by funswe** [fastify/fastify#1737]
I defined a routing that supports GET and POST methods. In scheme, querystring and body are both defined, but the body is also validated when the request method is GET.

then document writes : 
![image](https://user-images.githubusercontent.com/50731295/60958944-e4e22500-a339-11e9-849b-2c6a5e2e332e.png)

From the documentation point of view, I don't think the body is validated when the request is a GET method, but it's not.

```
fastify.route({
    method: ['GET', 'POST'],
    url: '/a/b',
    schema: {
        querystring: {
            type: 'object',
            properties: {
                name: {type: 'string'},
            }
        },
        body: {
            type: 'object',
            properties: {
                name: {type: 'string'},
            }
        }
    }
});
```

**Comment by mcollina** [comment]
I'm not sure I understand fully. Would you mind adding a self-contained example that we can run to trigger this behavior?

**Comment by funswe** [comment]
```
const fastify = require('fastify')({
    logger: true
});

fastify.route({
    method: ['GET', 'POST'],
    url: '/test',
    schema: {
        querystring: {
            type: 'object',
            properties: {
                name: {type: 'string'},
            }
        },
        body: {
            type: 'object',
            properties: {
                name: {type: 'string'},
            }
        }
    },
    handler: async () => {
        return 'hello';
    }
});

fastify.listen(3000, (err, address) => {
    if (err) throw err
    fastify.log.info(`server listening on ${address}`)
});
```

When I request using the GET method, return  “{"statusCode":400,"error":"Bad Request","message":"body should be object"}”

**Comment by funswe** [comment]
In my opinion, GET requests should not validate body

**Comment by funswe** [comment]
In my opinion, GET requests should not validate body 
![image](https://user-images.githubusercontent.com/50731295/60962041-f7f7f380-a33f-11e9-821f-7af7bbcac469.png)

**Comment by mcollina** [comment]
Are you sending a GET request with a body? What's the client for that server?

**Comment by funswe** [comment]
Because I'm not sure whether the client sends GET or POST requests, I add querystring and body to the scheme. What I hope is that whether GET or POST, the parameter `name` can be validated.

**Comment by mcollina** [comment]
Would you mind adding a client written in Node.js that demonstrate the error you are seeing, and what would be your expected behavior?

**Comment by RafaelGSS** [comment]
I think that understand him @mcollina. In doc the "body" just validate POST and PUT request, but if you try make GET request on this schema, they provide the error above.

**Comment by mcollina** [comment]
I think we should just throw an error in that case, adding a `body` schema to a GET route should not be valid.

**Comment by RafaelGSS** [comment]
Sure, but in these cases, create a route that accepts GET and POST should be thrown an error too?

**Comment by mcollina** [comment]
> Sure, but in these cases, create a route that accepts GET and POST should be thrown an error too?

I wouldn't do that. It's mostly related to the `body` validation that cannot be applied to `GET`. We might want to write a wildcard route that just forward the request to another service for example.

**Comment by Eomm** [comment]
I think that this behaviour should be chosen by the programmer that is creating the API because someone accepts a body in a GET request as [elasticsearch does](https://www.elastic.co/guide/en/elasticsearch/guide/current/_empty_search.html) referring to the RFC that doesn't deny to parse the body explicitly.

This use case could be just implemented two different routes that use the same handler or using the [fastify-schema-constraint](https://github.com/Eomm/fastify-schema-constraint) plugins that add a check in order to apply or not the schema validation if the request is GET or POST.

**Comment by delvedor** [comment]
I think that we should not allow defining a body in a GET route.
Elasticsearch does that for compatibility reason with legacy code, and I would not use it as an example.

Furthermore, you want to use the validation in a route so you can safely access the data afterward. If you declare a route with both GET and POST you will miss that, and you will encounter a bug in your code eventually.

**Comment by RafaelGSS** [comment]
I'll make the PR for throw a error when pass `schema.body` inside `GET` requests... We need it?

If yes: :+1:
If no: :-1:

**Comment by funswe** [comment]
> I'll make the PR for throw a error when pass `schema.body` inside `GET` requests... We need it?
> 
> If yes: 
> If no:

I think it's better to ignore ```schema. body``` when GET request.

**Comment by delvedor** [comment]
The problem with ignoring is that is not explicit.
There are two solutions, log a warn or throw an exception. I'm fine with both of them.

**Comment by RafaelGSS** [comment]
If we go ignore the `schema.body` when GET request, I think that throw an exception is good. For two reasons:
- We go to minimize future issues.
- Don't need to think about how to ignore it in future releases.

**Comment by mcollina** [comment]
Would you like to send a PR about throwing when `GET` route is defined with a `body` schema? I'll do this in fastify v3, to avoid any possible breakages.

**Comment by oxyii** [comment]
I think fastify can support GET body... e.g. for elasticsearch REST implementation https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-body.html

**Comment by mcollina** [comment]
Interesting.. but the above stands. I’d be ok to put it behind an option if somebody would need this.
Il 2 dic 2019, 03:07 +0100, Ruslan Baybekov <notifications@github.com>, ha scritto:

**Comment by oxyii** [comment]
@mcollina Unfortunately already [replaced](https://github.com/oxyii/mongoosastic-rest#post-dbcollection_aggregate) with `POST`

**Comment by climba03003** [comment]
Resolved by #3274 which target to `4.x`
